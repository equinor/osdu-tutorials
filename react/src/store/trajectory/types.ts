import { FetchTrajectoryPointsResponse } from 'api/trajectory.api';

export const FETCH_TRAJECTORY_POINTS_START = 'FETCH_TRAJECTORY_POINTS_START';
export const FETCH_TRAJECTORY_POINTS_SUCCESS = 'FETCH_TRAJECTORY_POINTS_SUCCESS';
export const FETCH_TRAJECTORY_POINTS_FAIL = 'FETCH_TRAJECTORY_POINTS_FAIL';

export const UNSELECT_TRAJECTORY = 'UNSELECT_TRAJECTORY';
export const UNSELECT_ALL_TRAJECTORIES = 'UNSELECT_ALL_TRAJECTORIES';

export interface FetchTrajectoryPointsStartAction {
  type: typeof FETCH_TRAJECTORY_POINTS_START;
  payload: {
    wellId: string;
    wellboreId: string;
  };
}

export interface FetchTrajectoryPointsSuccessAction {
  type: typeof FETCH_TRAJECTORY_POINTS_SUCCESS;
  payload: {
    wellId: string,
    result: FetchTrajectoryPointsResponse,
  };
}

export interface FetchTrajectoryPointsFailAction {
  type: typeof FETCH_TRAJECTORY_POINTS_FAIL;
  payload: {
    err: Error;
    wellboreId: string;
  };
}

export interface UnselectTrajectoryAction {
  type: typeof UNSELECT_TRAJECTORY;
  payload: string;
}

export interface UnselectAllTrajectoriesAction {
  type: typeof UNSELECT_ALL_TRAJECTORIES;
}

export type TrajectoryActionTypes =
  | FetchTrajectoryPointsStartAction
  | FetchTrajectoryPointsSuccessAction
  | FetchTrajectoryPointsFailAction
  | UnselectTrajectoryAction
  | UnselectAllTrajectoriesAction;
