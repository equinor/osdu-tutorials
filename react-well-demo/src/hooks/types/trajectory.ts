export type Trajectory = {
  points: WellboreTrajectoryPoint[];
  data: WellboreTrajectoryData;
};

type WellboreTrajectoryData = {
  ExtensionProperties: ExtensionProperties;
};

type ExtensionProperties = {
  trajectory: WellboreTrajectoryPoint[];
};

export type WellboreTrajectoryPoint = {
  md: number;
  tvd: number;
  azimuth: number;
  inclination: number;
  longitude: number;
  latitude: number;
};
