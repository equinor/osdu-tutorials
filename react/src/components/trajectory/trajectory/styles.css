.trajectory {
  width: 100%;
  height: 100%;
}

.trajectory-chart__layout {
  height: 100%;
}
