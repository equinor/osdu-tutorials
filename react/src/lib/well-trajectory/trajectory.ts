import { Vector3, VertexColors, WebGLRenderer } from 'three';
import { Line2 } from 'three/examples/jsm/lines/Line2';
import { LineGeometry } from 'three/examples/jsm/lines/LineGeometry';
import { LineMaterial } from 'three/examples/jsm/lines/LineMaterial';


interface TrajectoryDataPoint {
  measuredDepth: number;
  inclination: number;
  azimuth: number;
  x: number,
  y: number
}

interface TrajectoryData {
  points: TrajectoryDataPoint[];
}


export function calculateNextTrajectoryPoint(
  prev: Vector3,
  prevPointMeasuredDepth: number,
  nextPointData: TrajectoryDataPoint
): Vector3 {
  const nextPoint = new Vector3();
  const degToRad = Math.PI / 180;

  const depthDelta = nextPointData.measuredDepth - prevPointMeasuredDepth;
  const dx =
    depthDelta *
    Math.sin(degToRad * nextPointData.inclination) *
    Math.cos(degToRad * nextPointData.azimuth);
  const dy = -depthDelta * Math.cos(degToRad * nextPointData.inclination);
  const dz =
    depthDelta *
    Math.sin(degToRad * nextPointData.inclination) *
    Math.sin(degToRad * nextPointData.azimuth);

  nextPoint.x = prev.x + dx; //width and north direction
  nextPoint.y = prev.y + dy; //depth
  nextPoint.z = prev.z + dz; //height

  return nextPoint;
}

export function calculateTrajectoryPointsIn3D(trajectory: TrajectoryData): Vector3[] {
  const pointsIn3D: Vector3[] = [];
  const numberOfPoints: number = trajectory.points.length;

  const firstPointData = trajectory.points[0];
  const firstPoint = new Vector3(0, -firstPointData.measuredDepth, 0);
  pointsIn3D.push(firstPoint);

  let prevPoint = firstPoint.clone();
  let prevPointMeasuredDepth = firstPointData.measuredDepth;
  for (let i = 1; i < numberOfPoints; i++) {
    const nextPointData = trajectory.points[i];

    const nextPoint = calculateNextTrajectoryPoint(
      prevPoint,
      prevPointMeasuredDepth,
      nextPointData
    );
    pointsIn3D.push(nextPoint);

    prevPoint = nextPoint;
    prevPointMeasuredDepth = nextPointData.measuredDepth;
  }

  return pointsIn3D;
}

const lineColors = [
  [255, 255, 0],
  [255, 0, 255],
  [0, 255, 255],
  [255, 0, 0],
  [0, 255, 0],
  [0, 0, 255],
];

export function createLine2(points: Vector3[], lineIndex: number): Line2 {
  const positions: number[] = [];
  const colors: number[] = [];

  points.forEach(v => {
    positions.push(v.x, v.y, v.z);
    colors.push(...lineColors[lineIndex % lineColors.length]);
  });

  const lineGeometry = new LineGeometry();
  lineGeometry.setPositions(positions);
  lineGeometry.setColors(colors);

  const lineMaterial = new LineMaterial({
    color: 0xffffff,
    linewidth: 3,
    vertexColors: VertexColors,
    dashed: false
  });

  const line = new Line2(lineGeometry, lineMaterial);

  line.onBeforeRender = (renderer: WebGLRenderer) => {
    renderer.getSize(lineMaterial.resolution)
  }

  return line;
}

export function createTrajectoryLine(points: Vector3[], lineIndex: number) {
  const trajectory = createLine2(points, lineIndex);

  const trajectoryFirstPointHeight = points[0].y;
  trajectory.translateY(-trajectoryFirstPointHeight);

  return trajectory;
}
