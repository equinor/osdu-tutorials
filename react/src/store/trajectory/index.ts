export {
  fetchTrajectoryPointsAction,
  unselectTrajectoryAction,
  unselectAllTrajectoriesAction,
} from './actions';
export * from './reducer';
