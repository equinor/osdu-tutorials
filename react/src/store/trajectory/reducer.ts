import {
  TrajectoryActionTypes,
  FETCH_TRAJECTORY_POINTS_SUCCESS,
  FETCH_TRAJECTORY_POINTS_FAIL,
  FETCH_TRAJECTORY_POINTS_START,
  UNSELECT_TRAJECTORY,
  UNSELECT_ALL_TRAJECTORIES,
} from './types';

export interface TrajectoryState {
  /** a current well, whose trajectories a to draw */
  wellId: string;

  /** a set of trajectories chosen by a user to visialize */
  trajectories: TrajectoryToDraw[];
}

const initialState: TrajectoryState = {
  wellId: '',
  trajectories: [],
};

export interface TrajectoryToDraw {
  wellboreId: string;

  /** trajectory data fetch api request progress flag */
  isLoading: boolean;

  /** proof that a fetch request was made */
  isLoaded: boolean;

  /** any error, occured during a trajectory data fetch */
  loadError?: Error;

  /** a set of points to build a trajectory's 3D model */
  points: TrajectoryPoint[];
}

interface TrajectoryPoint {
  measuredDepth: number;
  inclination: number;
  azimuth: number;
  x: number;
  y: number;
}

export const trajectoryReducer = (
  state: TrajectoryState = initialState,
  action: TrajectoryActionTypes
): TrajectoryState => {
  switch (action.type) {
    case FETCH_TRAJECTORY_POINTS_START:
      return {
        ...state,
        wellId: action.payload.wellId,
        trajectories: [
          ...(state.wellId === action.payload.wellId ? state.trajectories : []),
          {
            wellboreId: action.payload.wellboreId,
            isLoading: true,
            isLoaded: false,
            loadError: undefined,
            points: [],
          },
        ],
      };
    case FETCH_TRAJECTORY_POINTS_SUCCESS:
      return {
        ...state,
        trajectories: state.trajectories.map(t => {
          if (t.wellboreId !== action.payload.result.wellbore_id) {
            return t;
          }

          const data = action.payload.result.data;
          const points = data.azimuths.map(
            (a, i): TrajectoryPoint => ({
              azimuth: data.azimuths[i],
              inclination: data.inclinations[i],
              measuredDepth: data.measured_depths[i],
              x: data.surface_x[i],
              y: data.surface_y[i],
            })
          );

          return {
            ...t,
            isLoading: false,
            isLoaded: true,
            points,
          };
        }),
      };
    case FETCH_TRAJECTORY_POINTS_FAIL:
      return {
        ...state,
        trajectories: state.trajectories.map(t => {
          return action.payload.wellboreId === t.wellboreId
            ? {
              ...t,
              isLoading: false,
              isLoaded: true,
              loadError: action.payload.err,
            }
            : t;
        }),
      };
    case UNSELECT_TRAJECTORY:
      return {
        ...state,
        trajectories: state.trajectories.filter(t => action.payload !== t.wellboreId),
      };
    case UNSELECT_ALL_TRAJECTORIES:
      return {
        ...state,
        trajectories: [],
      };
    default:
      return state;
  }
};
