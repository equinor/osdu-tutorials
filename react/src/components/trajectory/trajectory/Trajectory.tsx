import React, { useRef, useEffect } from 'react';
import './styles.css';

import { createTrajectoryChart } from 'lib/well-trajectory';
import { TrajectoryToDraw } from 'store/trajectory';

export interface TrajectoryProps {
    trajectoriesToDraw: TrajectoryToDraw[];
}

/**
 * Draws a 3D models of a well's trajectory
 * multiple trajectories can be drawn, layering on each other
 * @param {TrajectoryProps} props
 */
export function Trajectory({ trajectoriesToDraw }: TrajectoryProps) {
    // threejs is not integrated into a react, it just needs a container to renter into
    const ref = useRef<HTMLDivElement>(null);

    useEffect(() => {
        if (ref.current) {
            createTrajectoryChart(ref.current, trajectoriesToDraw);
        }
    }, [trajectoriesToDraw]);

    return <div className="trajectory" ref={ref}></div>;
}