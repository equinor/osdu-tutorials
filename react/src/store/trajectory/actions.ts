import { AppState } from 'store';
import { ThunkAction } from 'redux-thunk';
import { fetchTrajectoryPoints, FetchTrajectoryPointsResponse } from 'api/trajectory.api';
import {
  TrajectoryActionTypes,
  FETCH_TRAJECTORY_POINTS_START,
  FETCH_TRAJECTORY_POINTS_FAIL,
  FETCH_TRAJECTORY_POINTS_SUCCESS,
  UNSELECT_TRAJECTORY,
  UNSELECT_ALL_TRAJECTORIES,
} from './types';

type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, null, TrajectoryActionTypes>;

/**
 * remove a trajectory from a list to draw
 * @param wellboreId
 */
export function unselectTrajectoryAction(wellboreId: string): TrajectoryActionTypes {
  return {
    type: UNSELECT_TRAJECTORY,
    payload: wellboreId,
  };
}

/** clear a canvas from every trajectory */
export function unselectAllTrajectoriesAction(): TrajectoryActionTypes {
  return {
    type: UNSELECT_ALL_TRAJECTORIES,
  };
}

export function fetchTrajectoryPointsStartAction(wellId: string, wellboreId: string): TrajectoryActionTypes {
  return {
    type: FETCH_TRAJECTORY_POINTS_START,
    payload: { wellId, wellboreId },
  };
}

export function fetchTrajectoryPointsSuccessAction(
  wellId: string,
  response: FetchTrajectoryPointsResponse
): TrajectoryActionTypes {
  return {
    type: FETCH_TRAJECTORY_POINTS_SUCCESS,
    payload: {
      wellId,
      result: response,
    },
  };
}

export function fetchTrajectoryPointsFailAction(err: Error, wellboreId: string): TrajectoryActionTypes {
  return {
    type: FETCH_TRAJECTORY_POINTS_FAIL,
    payload: {
      err,
      wellboreId,
    },
  };
}

/**
 * fetch from the backend a trajectory data
 * it is possible to draw multiple trajectories for a single well, therefore Well identifier is required
 * @param wellId
 * @param wellboreId
 */
export const fetchTrajectoryPointsAction = (wellId: string, wellboreId: string): AppThunk => dispatch => {
  console.log("fetching mock data");
  dispatch(fetchTrajectoryPointsStartAction(wellId, wellboreId));
  const mockResponse: FetchTrajectoryPointsResponse = {
    wellbore_id: wellboreId,
    data: {
      measured_depths: [109, 122.95, 132.98, 142.96, 152.93],
      azimuths: [0, 171.75, 251.31, 250.95, 231.63],
      inclinations: [0, 0.07, 0.08, 0.09, 0.04],
      surface_x: [58.43590277, 58.43590277, 58.43590277, 58.43590277, 58.43590277],
      surface_y: [1.929736075, 1.929736075, 1.929736075, 1.929736075, 1.929736075],
    }
  }
  return Promise.resolve(mockResponse).then(data => dispatch(fetchTrajectoryPointsSuccessAction(wellId, data)));
  /*
  dispatch(fetchTrajectoryPointsStartAction(wellId, wellboreId));

  return fetchTrajectoryPoints(wellboreId)
    .then(data => dispatch(fetchTrajectoryPointsSuccessAction(wellId, data)))
    .catch(err => dispatch(fetchTrajectoryPointsFailAction(err, wellboreId)));
  */
};
