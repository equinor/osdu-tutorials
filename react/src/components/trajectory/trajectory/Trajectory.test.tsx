import React from 'react';
import { shallow } from 'enzyme';
import { Trajectory, TrajectoryProps } from './Trajectory';
import { trajectoryFetchPointsSuccessState } from 'test-data/trajectory';

describe('Trajectory', () => {
  it('should render self and subcomponents', () => {
    const props: TrajectoryProps = {
      trajectoriesToDraw: trajectoryFetchPointsSuccessState.trajectories,
      className: 'test-class-name',
    };
    const wrapper = shallow(<Trajectory {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
