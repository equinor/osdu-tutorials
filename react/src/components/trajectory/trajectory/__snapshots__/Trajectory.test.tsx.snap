// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trajectory should render self and subcomponents 1`] = `
<div
  className="trajectory"
/>
`;
